infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuInterClusterReplicationServer.java

```diff
@@ -290,21 +304,84 @@ private ServerContext getServerContext(Map<String, Object> opts) {
+     * any ongoing rolling upgrade completes.
      *
      * @param serverContext server context (server information)
      */
     private void startDiscoveryService(ServerContext serverContext) {
 
+        CorfuRuntime runtime = getRuntime(serverContext);
+        CorfuStore corfuStore = new CorfuStore(runtime);
+
+        LogReplicationUpgradeManager upgradeManager =
+                new LogReplicationUpgradeManager(corfuStore, serverContext.getPluginConfigFilePath());
+
+        // Check if an upgrade is in progress.  If it is, wait for it to complete
+        log.info("Wait for any ongoing rolling upgrade to complete.");
+
+        while(isUpgradeInProgress(upgradeManager, corfuStore)) {
+            // Wait before checking again.
+            sleep();
+        }
+
+        // Upgrade has completed.  Start the discovery Service
         log.info("Start Discovery Service.");
 
         // Start LogReplicationDiscovery Service, responsible for
         // acquiring lock, retrieving Site Manager Info and processing this info
         // so this node is initialized as Source (sender) or Sink (receiver)
-        replicationDiscoveryService = new CorfuReplicationDiscoveryService(serverContext);
+        replicationDiscoveryService = new CorfuReplicationDiscoveryService(serverContext, runtime);
         replicationDiscoveryService.start();
     }
 
+    private CorfuRuntime getRuntime(ServerContext serverContext) {
+
+        String localCorfuEndpoint = getCorfuEndpoint(getHostFromEndpointURL(serverContext.getLocalEndpoint()),
+            serverContext.getCorfuServerConnectionPort());
+
+        return CorfuRuntime.fromParameters(CorfuRuntime.CorfuRuntimeParameters.builder()
+            .trustStore((String) serverContext.getServerConfig().get(ConfigParamNames.TRUST_STORE))
+            .tsPasswordFile((String) serverContext.getServerConfig().get(ConfigParamNames.TRUST_STORE_PASS_FILE))
+            .keyStore((String) serverContext.getServerConfig().get(ConfigParamNames.KEY_STORE))
+            .ksPasswordFile((String) serverContext.getServerConfig().get(ConfigParamNames.KEY_STORE_PASS_FILE))
+            .tlsEnabled((Boolean) serverContext.getServerConfig().get("--enable-tls"))
+            .systemDownHandler(() -> System.exit(SYSTEM_EXIT_ERROR_CODE))
+            .maxCacheEntries(serverContext.getLogReplicationCacheMaxSize()/2)
+            .maxWriteSize(serverContext.getMaxWriteSize())
+            .build())
+            .parseConfigurationString(localCorfuEndpoint).connect();
+    }
+
+    private String getCorfuEndpoint(String localHostAddress, int port) {
+        return getVersionFormattedEndpointURL(localHostAddress, port);
+    }
+
+    private boolean isUpgradeInProgress(LogReplicationUpgradeManager upgradeManager, CorfuStore corfuStore) {
+        boolean isUpgraded = true;
+        try (TxnContext txnContext = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
```

Comment:
should we have this in a retry block? 
isLrUpgradeInProgress writes to the event table when all the nodes are migrated...we should retry if the write fails, right?

Reply:
good point.  I had started writing an IRetry block for it but there were a few limitations due to which it could not be used:
1. This block must run for a long time(potentially forever) until the upgrade completes.  If an IRetry thread is interrupted, we usually throw an UnrecoverableCorfuInterrupted error.  In this case, it is incorrect.
2. The default retry interval is 1 second which I felt was quite short and can cause multiple repeated checks.  Did not find a way to change this parameter.

We do take the concurrent writes in account and catch this exception and retry in the current implementation.